File: ex_5_fifo.cpp
#include<iostream>
#include<vector>
#include<algorithm>
using namespace std;

class fifo_page_replacement{
    private:
    int frame_count;
    vector<int> frames;

    public:
    fifo_page_replacement(int count): frame_count(count){}

    bool page_fault_occurred(int page){
        if (find(frames.begin(),frames.end(),page)==frames.end()){
            if (frames.size() < frame_count){
                frames.push_back(page);
                return true;  // Page fault occurred
            }
            else{
                frames.erase(frames.begin());
                frames.push_back(page);
                return true;  // Page fault occurred
            }
        }
        return false;  // No page fault occurred
    }

    void print_frames(){
        cout<<"Current frames ";
        for (int frame: frames){
            cout<<frame<<" ";
        }
        cout<<endl;
    }
};

int main(){
    fifo_page_replacement fifo(3);
    vector<int> pages = {1, 2, 3, 4, 1, 2, 5, 1, 2, 3, 4};
    int page_faults = 0;

    for (int page: pages){
        if (fifo.page_fault_occurred(page)) {
            page_faults++;
        }
        fifo.print_frames();
    }

    cout << "Total number of page faults: " << page_faults << endl;

    return 0;
}

File: ex_5_fifo.py
class FIFOPageReplacement:
    def __init__(self, frame_count):
        self.frame_count = frame_count
        self.frames = []

    def page_fault_occurred(self, page):
        if page not in self.frames:
            if len(self.frames) < self.frame_count:
                self.frames.append(page)
                return True  # Page fault occurred
            else:
                self.frames.pop(0)
                self.frames.append(page)
                return True  # Page fault occurred
        return False  # No page fault occurred

    def print_frames(self):
        print("Current frames: ", self.frames)


fifo = FIFOPageReplacement(3)
pages = [1, 2, 3, 4, 1, 2, 5, 1, 2, 3, 4, 5]
page_faults = 0

for page in pages:
    if fifo.page_fault_occurred(page):
        page_faults += 1
    fifo.print_frames()

print("Total number of page faults:", page_faults)
